### PR TITLE
Interactions refactoring

### LINKED ISSUE #1056: Granular interactivity configuration
At the moment, there is only one option that can be passed to the map: interactive (true/false), which disable every interaction. It would be good to have different options:
 - rotation
 - draggable
 - boxzoom
 - tilt
 - ... other future options

It would be better to have something similar to leaflet "handlers":
 - Being able to pass options to the map corresponding to each interactivity so that the can be disabled by default
 - Being able to access the handler's instance and call enable() / disable() on the fly to activate an interaction on demand

This is a follow-up on the comments of @jfirebaugh and @mourner on issue #1038:
https://github.com/mapbox/mapbox-gl-js/issues/1038#issuecomment-77328219

Fabien

### PR DESCRIPTION
**Pull request by mourner**
Closes #1056. A work in progress.

Makes each interaction handler an independent class, with code consolidated in one place and 10 times easier to understand, and ability to enable/disable handlers individually.

- [x] scroll zoom
- [x] box zoom
- [x] drag rotate
- [x] drag pan
- [x] pinch zoom
- [x] double click/tap
- [x] keyboard
- [x] map events
- [x] clean up
- [ ] add docs where undocumented

### CODE REVIEW

**Comment by jfirebaugh** [comment]
:ok_hand:

**Comment by mourner** [comment]
@jfirebaugh OK, this is ready for review. 
All commits are self-contained, working and passing tests so no need to squash.

**Comment by jfirebaugh** [comment]
This is fantastic, please merge.

File: js/ui/handler/dblclick_zoom.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+'use strict';
+
+module.exports = DragPan;
```
Comment (jfirebaugh) [inline js/ui/handler/dblclick_zoom.js]: `s/DragPan/DblclickZoom/g`

**Pull request merged**